tracing-appender/src/rolling.rs

```diff
@@ -253,21 +227,37 @@ impl RollingFileAppender {
 
 impl io::Write for RollingFileAppender {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let should_rollover = |now: OffsetDateTime| -> bool {
+            if let Some(current_time) = self.state.should_rollover_by_date(now) {
+                let _did_cas = self.state.advance_date(now, current_time);
+                debug_assert!(_did_cas, "if we have &mut access to the appender, no other thread can have advanced the timestamp...");
+                return true;
+            }
+            if self.state.should_rollover_by_size() {
+                self.state.advance_same_date_filename_index();
+                return true;
+            }
+            false
+        };
+
         let now = self.now();
+        let should_rollover = should_rollover(now);
+
         let writer = self.writer.get_mut();
-        if let Some(current_time) = self.state.should_rollover(now) {
-            let did_cas = self.state.advance_date_and_index(now, current_time);
-            debug_assert!(did_cas, "if we have &mut access to the appender, no other thread can have advanced the timestamp...");
+        if should_rollover {
             self.state.refresh_writer(now, writer);
         }
 
         let written = writer.write(buf)?;
-
         self.state.current_size.fetch_add(
-            u64::try_from(written).expect("usize to u64 conversion"),
+            u64::try_from(written).map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::Other,
+                    format!("Failed to convert usize to u64: {}", e),
```

Comment:
nitpick: Error messages should be formatted using a single sentence which:

- is short and concise,
- does not start by a capital letter, and
- does not contain trailing punctuation.

This is the convention adopted by the Rust project:

- [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#error-types-are-meaningful-and-well-behaved-c-good-err)
- [std::error::Error](https://doc.rust-lang.org/stable/std/error/trait.Error.html)

rationale: It’s easier to compose with other error messages.

```suggestion
                    format!("failed to convert usize to u64: {}", e),
```

Reply:
Nice, this settles an internal battle I've had for a long time now 😆

Reply:
Shouldn't be also safe to unwrap the usize/u64 conversion here?

Reply:
Yeah, I think it’s fine!